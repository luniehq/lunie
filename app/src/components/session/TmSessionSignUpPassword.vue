<template>
  <SessionFrame>
    <TmFormStruct :submit="onSubmit">
      <h2 class="session-title">Choose a password</h2>
      <div>
        <Steps
          v-if="!session.mobile"
          :steps="[`Create`, `Password`, `Backup`]"
          active-step="Password"
        />
        <TmFormGroup
          :error="$v.fieldPassword.$error"
          field-id="sign-up-password"
          field-label="Password"
        >
          <TmField
            id="sign-up-password"
            v-model="fieldPassword"
            v-focus
            type="password"
            placeholder="Must be at least 10 characters"
          />
          <TmFormMsg
            v-if="$v.fieldPassword.$error && !$v.fieldPassword.required"
            name="Password"
            type="required"
          />
          <TmFormMsg
            v-if="$v.fieldPassword.$error && !$v.fieldPassword.minLength"
            name="Password"
            type="minLength"
            min="10"
          />
        </TmFormGroup>
        <TmFormGroup
          :error="$v.fieldPasswordConfirm.$error"
          field-id="sign-up-password-confirm"
          field-label="Confirm Password"
        >
          <TmField
            id="sign-up-password-confirm"
            v-model="fieldPasswordConfirm"
            type="password"
            placeholder="Enter password again"
          />
          <TmFormMsg
            v-if="
              $v.fieldPasswordConfirm.$error &&
              !$v.fieldPasswordConfirm.sameAsPassword
            "
            name="Password confirmation"
            type="match"
          />
        </TmFormGroup>
        <div class="session-footer">
          <TmBtn value="Next" type="submit" />
        </div>
      </div>
    </TmFormStruct>
  </SessionFrame>
</template>

<script>
import { mapState } from "vuex"
import { required, minLength, sameAs } from "vuelidate/lib/validators"
import TmBtn from "common/TmBtn"
import TmFormGroup from "common/TmFormGroup"
import TmFormStruct from "common/TmFormStruct"
import TmField from "common/TmField"
import TmFormMsg from "common/TmFormMsg"
import SessionFrame from "common/SessionFrame"
import Steps from "../../ActionModal/components/Steps"

export default {
  name: `session-sign-up`,
  components: {
    TmBtn,
    TmField,
    SessionFrame,
    TmFormGroup,
    TmFormMsg,
    TmFormStruct,
    Steps,
  },
  computed: {
    ...mapState([`session`, `signup`]),
    fieldPassword: {
      get() {
        return this.$store.state.signup.signUpPassword
      },
      set(value) {
        this.$store.commit(`updateField`, { field: `signUpPassword`, value })
      },
    },
    fieldPasswordConfirm: {
      get() {
        return this.$store.state.signup.signUpPasswordConfirm
      },
      set(value) {
        this.$store.commit(`updateField`, {
          field: `signUpPasswordConfirm`,
          value,
        })
      },
    },
  },
  methods: {
    async onSubmit() {
      this.$v.$touch()
      if (this.$v.$error) return
      this.$router.push(`/create/confirm`)
    },
  },
  validations: () => ({
    fieldPassword: { required, minLength: minLength(10) },
    fieldPasswordConfirm: { sameAsPassword: sameAs(`fieldPassword`) },
  }),
}
</script>
