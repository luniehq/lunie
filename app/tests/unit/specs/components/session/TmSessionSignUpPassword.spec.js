import { createLocalVue, shallowMount } from "@vue/test-utils"
import Vuelidate from "vuelidate"
import TmSessionSignUpPassword from "session/TmSessionSignUpPassword"

describe(`TmSessionSignUpPassword`, () => {
  const localVue = createLocalVue()

  localVue.directive(`focus`, () => {})
  localVue.use(Vuelidate)

  let wrapper, $store

  beforeEach(() => {
    $store = {
      state: {
        session: { insecureMode: true },
        signup: {
          signUpPassword: ``,
          signUpPasswordConfirm: ``,
        },
      },
      commit: jest.fn(),
      dispatch: jest.fn(),
      mutations: {
        updateField: jest.fn(),
      },
    }

    wrapper = shallowMount(TmSessionSignUpPassword, {
      localVue,
      mocks: {
        $store,
        $router: {
          push: jest.fn(),
        },
      },
      stubs: [`router-link`],
      sync: false,
    })
  })

  it("renders", () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it(`should commit updateField on password field change`, () => {
    wrapper.setData({ fieldPassword: `1234567890` })
    expect($store.commit).toHaveBeenCalledWith(`updateField`, {
      field: `signUpPassword`,
      value: `1234567890`,
    })
  })

  it(`should commit updateField on password confirm field change`, () => {
    wrapper.setData({ fieldPasswordConfirm: `1234567890` })
    expect($store.commit).toHaveBeenCalledWith(`updateField`, {
      field: `signUpPasswordConfirm`,
      value: `1234567890`,
    })
  })

  it(`should go to /create/confirm when submit the form`, async () => {
    wrapper.vm.$store.state.signup.signUpPassword = `1234567890`
    wrapper.vm.$store.state.signup.signUpPasswordConfirm = `1234567890`
    await wrapper.vm.onSubmit()
    expect(wrapper.vm.$router.push).toHaveBeenCalledWith(`/create/confirm`)
  })
})
