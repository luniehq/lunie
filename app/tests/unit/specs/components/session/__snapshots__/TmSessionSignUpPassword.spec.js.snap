// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TmSessionSignUpPassword renders 1`] = `
<sessionframe-stub
  icon=""
>
  <tmformstruct-stub
    submit="function () { [native code] }"
    width=""
  >
    <h2
      class="session-title"
    >
      Choose a password
    </h2>
     
    <div>
      <steps-stub
        activestep="Password"
        steps="Create,Password,Backup"
      />
       
      <tmformgroup-stub
        fieldid="sign-up-password"
        fieldlabel="Password"
      >
        <tmfield-stub
          id="sign-up-password"
          placeholder="Must be at least 10 characters"
          type="password"
          value=""
        />
         
        <!---->
         
        <!---->
      </tmformgroup-stub>
       
      <tmformgroup-stub
        fieldid="sign-up-password-confirm"
        fieldlabel="Confirm Password"
      >
        <tmfield-stub
          id="sign-up-password-confirm"
          placeholder="Enter password again"
          type="password"
          value=""
        />
         
        <!---->
      </tmformgroup-stub>
       
      <div
        class="session-footer"
      >
        <tmbtn-stub
          customclass=""
          type="submit"
          value="Next"
        />
      </div>
    </div>
  </tmformstruct-stub>
</sessionframe-stub>
`;
